File: Circle Detection.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import org.opencv.core.Core;
import org.opencv.core.Mat;
import org.opencv.core.Point;
import org.opencv.core.Scalar;
import org.opencv.highgui.HighGui;
import org.opencv.imgcodecs.Imgcodecs;
import org.opencv.imgproc.Imgproc;

public class CircleDetector {
    private String image_path;
    private int population_size;
    private int max_generations;
    private Mat img;
    private int height, width;
    private List<int[]> population;
    private double best_fitness = Double.POSITIVE_INFINITY;
    private int[] best_individual = null;
    private List<Double> fitness_values = new ArrayList<Double>();

    public CircleDetector(String image_path, int population_size, int max_generations) {
        this.image_path = image_path;
        this.population_size = population_size;
        this.max_generations = max_generations;
        this.img = Imgcodecs.imread(this.image_path, Imgcodecs.IMREAD_GRAYSCALE);
        this.height = this.img.height();
        this.width = this.img.width();
        this.population = new ArrayList<int[]>();
    }

    public void initialize_population() {
        Random rand = new Random();
        for (int i = 0; i < this.population_size; i++) {
            int[] individual = new int[3];
            int x = rand.nextInt(this.width);
            int y = rand.nextInt(this.height);
            int r = rand.nextInt(Math.min(this.width, this.height)/2) + 1;
            individual[0] = x;
            individual[1] = y;
            individual[2] = r;
            this.population.add(individual);
        }
    }

    public double evaluate_fitness(int[] individual) {
        double fitness = 0;
        for (int i = 0; i < this.width; i++) {
            for (int j = 0; j < this.height; j++) {
                int x = individual[0];
                int y = individual[1];
                int r = individual[2];
                double distance = Math.sqrt(Math.pow(i - x, 2) + Math.pow(j - y, 2));
                fitness += Math.abs(distance - r - this.img.get(j, i)[0]);
            }
        }
        return fitness;
    }

    public List<int[]> selection() {
        double fitness_sum = 0;
        for (Double fitness : this.fitness_values) {
            fitness_sum += fitness;
        }
        List<Double> probabilities = new ArrayList<Double>();
        for (Double fitness : this.fitness_values) {
            probabilities.add(fitness/fitness_sum);
        }
        List<Integer> selected_indices = new ArrayList<Integer>();
        while (selected_indices.size() < this.population_size/2) {
            double rand = Math.random();
            double cumulative_probability = 0;
            for (int i = 0; i < probabilities.size(); i++) {
                cumulative_probability += probabilities.get(i);
                if (rand <= cumulative_probability) {
                    selected_indices.add(i);
                    break;
                }
            }
        }
        List<int[]> selected_individuals = new ArrayList<int[]>();
        for (int i = 0; i < selected_indices.size(); i++) {
            selected_individuals.add(this.population.get(selected_indices.get(i)));
        }
        return selected_individuals;
    }

    public int[][] crossover(int[] parent1, int[] parent2) {
        int[][] children = new int[2][3];
        Random rand = new Random();
        int crossover_point = rand.nextInt(3);
        for (int i = 0;

File: Circle Detection.py
import cv2
import numpy as np
import random

class CircleDetector:
    def __init__(self, image_path, population_size, max_generations):
        self.image_path = image_path
        self.population_size = population_size
        self.max_generations = max_generations
        self.img = cv2.imread(self.image_path, 0)
        self.height, self.width = self.img.shape
        self.population = []
        self.best_fitness = float('inf')
        self.best_individual = None
        self.fitness_values = []

    def initialize_population(self):
        for i in range(self.population_size):
            individual = []
            x = random.randint(0, self.width)
            y = random.randint(0, self.height)
            r = random.randint(1, min(self.width, self.height)//2)
            individual.append(x)
            individual.append(y)
            individual.append(r)
            self.population.append(individual)

    def evaluate_fitness(self, individual):
        fitness = 0
        for i in range(self.width):
            for j in range(self.height):
                x, y, r = individual
                distance = np.sqrt((i - x)**2 + (j - y)**2)
                fitness += abs(distance - r - self.img[i][j])
        return fitness

    def selection(self):
        fitness_sum = sum(self.fitness_values)
        probabilities = [fitness/fitness_sum for fitness in self.fitness_values]
        selected_indices = np.random.choice(self.population_size, self.population_size//2, replace=False, p=probabilities)
        selected_individuals = [self.population[i] for i in selected_indices]
        return selected_individuals

    def crossover(self, parent1, parent2):
        child1 = []
        child2 = []
        crossover_point = random.randint(0, 2)
        for i in range(crossover_point):
            child1.append(parent1[i])
            child2.append(parent2[i])
        for i in range(crossover_point, 3):
            child1.append(parent2[i])
            child2.append(parent1[i])
        return child1, child2

    def mutation(self, individual):
        mutation_rate = 0.1
        for i in range(3):
            if random.random() < mutation_rate:
                if i == 0:
                    individual[i] = random.randint(0, self.width)
                elif i == 1:
                    individual[i] = random.randint(0, self.height)
                else:
                    individual[i] = random.randint(1, min(self.width, self.height)//2)
        return individual

    def evolve(self):
        self.initialize_population()
        for generation in range(self.max_generations):
            self.fitness_values = [self.evaluate_fitness(individual) for individual in self.population]
            if min(self.fitness_values) < self.best_fitness:
                self.best_fitness = min(self.fitness_values)
                self.best_individual = self.population[self.fitness_values.index(self.best_fitness)]
            selected_individuals = self.selection()
            self.population = []
            for i in range(self.population_size//2):
                parent1 = selected_individuals[i]
                parent2 = selected_individuals[-i-1]
                child1, child2 = self.crossover(parent1, parent2)
                child1 = self.mutation(child1)
                child2 = self.mutation(child2)
                self.population.append(child1)
                self.population.append(child2)
        return self.best_individual

if __name__ == "__main__":
    detector = CircleDetector("/home/rakib/Documents/4th Kibo RPC/circle.png", 50, 100)
    best_individual = detector.evaluate_fitness
    cv2.imshow('Circle Detected',img)
    cv2.waitKey(0)
    cv2.destroyAllWindows()
